### ISSUE TITLE
debug: option to disable the modal dialog on launch failure

### ISSUE THREAD

**Issue by hyangah** [microsoft/vscode#128484]
This is a resurrection of the original feature request https://github.com/microsoft/vscode/issues/48606
(Previously discussed in https://github.com/microsoft/vscode-go/issues/1644 in microsoft/vscode-go,
but we are still getting the same requests from users https://github.com/golang/vscode-go/issues/1375) 

The feature request https://github.com/microsoft/vscode-go/issues/1644 filed in microsoft/vscode-go
repo was rejected, I am afraid there was some misunderstanding during the discussion.

Go's debug adapter&debugger (`delve dap`) does not only launch the debugee upon a launch request,
but *also* performs the compile/build of the target binary.  That is different from debug workflows of 
other languages that usually separate build steps from debug launch steps. 
When build fails, the adapter returns error for the launch request
(we couldn't build due to compiler errors so we couldn't launch!), which results in the modal popup.

Revisiting @weinand's original comment https://github.com/microsoft/vscode/issues/48606#issuecomment-384265967

> I think there are two types of notifications:
>
>  - expected (frequent) problems like compiler errors,
>  - unexpected (rare) problems like a runtime error in a debugger or a non-existing variable in a launch config.
>
> IMO the first type of problem should be addressed by the non-modal notification whereas the second type
should result in a modal dialog because it is a fundamental problem that requires the users attention.

So, I'd say Go's debug extension current behavior doesn't match the expected UX.

Is there any workaround in DAP to implement this in vscode-go side?
Separating the compile step from the launch as a pre-launch task is not ideal for Go
since we depend on Delve's decision on how to build the target with sufficient debug info.

Can we have an option to disable the modal dialog depending on the types of launch failures?

cc @polinasok @suzmue @isidorn 

ps. I tried to use the `showUser` property (set to false), but that just controls the extra notification popup currently.

**Comment by polinasok** [comment]
Here is a sample log of the request-response traffic that results in the dialogue.
```
From client: initialize({"clientID":"vscode","clientName":"Visual Studio Code","adapterID":"go","pathFormat":"path","linesStartAt1":true,"columnsStartAt1":true,"supportsVariableType":true,"supportsVariablePaging":true,"supportsRunInTerminalRequest":true,"locale":"en-us","supportsProgressReporting":true,"supportsInvalidatedEvent":true})
To client: {"seq":0,"type":"response","request_seq":1,"command":"initialize","success":true,"body":{"supportsConditionalBreakpoints":true,"supportsConfigurationDoneRequest":true,"supportsSetVariable":true}}
From client: launch({"name":"Launch file","type":"go","request":"launch","program":"/Users/polina/go/src/temp.go,...}}
To client: {"seq":0,"type":"response","request_seq":2,"command":"launch","success":false,"message":"Failed to continue: Check the debug console for details.","body":{"error":{"id":3000,"format":"Failed to continue: Check the debug console for details.","showUser":false}}}
```

I guess we could return launch success and then immediately send a terminated event while printing the error details to the console to avoid the pop-up, but that seems like a hack.

I think it would be much cleaner for vscode to either honor `showUser` or to have a global setting to turn these pop-ups on and off.

**Comment by polinasok** [comment]
@weinand Any updates? We have a [draft PR](https://github.com/go-delve/delve/pull/2617), but it is too hacky to put inside of the delve debugger and negatively impacts clients that do not have this problem. Next step would be to move the hack into the vscode-go extension, intercepting the launch error response and replacing it with a success response + terminated event to trick vscode editor into not showing the window, but we would much rather avoid any such hacks if a mechanism is coming on your end soon. Thank you.

**Comment by weinand** [comment]
I think it is not a good idea to run the compile/build as part of DAP's launch request (and not as a separate task). Communicating compile/build errors to the user via the debug console is not a good user experience. I would expect that compile/build problems show up in the problems view and I can easily navigate from there.

...

**Comment by isidorn** [comment]
> VS Code only shows a modal dialog if the "showUser" property of the error message exists and is false. If "showUser" is missing, we would behave like before in order to preserve compatibility.

I think this would work and would not be a large change.
For reference here's a place where I think we show the modal dialog https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/debug/browser/debugService.ts#L836

**Comment by hyangah** [comment]
> I think it is not a good idea to run the compile/build as part of DAP's launch request (and not as a separate task). Communicating compile/build errors to the user via the debug console is not a good user experience. I would expect that compile/build problems show up in the problems view and I can easily navigate from there.

The build diagnostics show up in the problems view almost real time (thank to the language server) but I observed many users prefer continuous testing rather than switching among the problems view, debug console, and terminal.

> Debugging compiled languages always depends on compile/builds with sufficient debug info. But the C++ or C# extension compile/build as a prelaunch task and not as part of DAP launch and they never complained about this.

Maybe it's because go's tool chain is generally much easier to use so requiring a separate make file or configuration is not part of the common go dev workflow. For example, `dlv test` is modeled after `go test` command that takes either file path, directory path, or even package path, builds the test binary, and runs the test binary from the package directory. No more than one command is needed. In addition to what `go test` does, `dlv test` detects the `go` version user in use and provides the additional flags to include additional debug info which may vary depending on the go version. We may choose to let the extension duplicate what dlv would do and provide a pretask, but I think it's better to keep all this knowledge centralized in dlv and debug adapter (for vscode-go and other editors' plugins)

>> VS Code only shows a modal dialog if the "showUser" property of the error message exists and is false. If "showUser" is missing, we would behave like before in order to preserve compatibility.
> I think this would work and would not be a large change.
> For reference here's a place where I think we show the modal dialog https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/debug/browser/debugService.ts#L836

It sounds like the best option for us :-)

**Comment by weinand** [comment]
@isidorn ok, let's do this.

**Comment by polinasok** [comment]
https://github.com/go-delve/delve/pull/2617#issuecomment-885745255

> I think it is not a good idea to run the compile/build as part of DAP's launch request (and not as a separate task).

If this is part of the DAP contract or best practices, it should be documented in the spec. Otherwise, it really does seem quite logical to do whatever steps are needed, including compilation, as part of launching the debug session.

> The DAP "launch" request should fail without a modal dialog if the reason was a compile/build error.
> 
> I see two options:
> 
> * in case of a compile/build error in the "launch" request, the Go DA sends the error information to the debug console but does not return an error from the "launch" request to make VS Code continue the session normally. After the initialisation sequence is done, the Go DA sends a "Terminated" and "Exited" events to stop the debug session. This would not require any changes in clients (e.g. VS Code).

This is what [this PR](https://github.com/go-delve/delve/pull/2617) has attempted to do. It looked wrong to report success when we had a clear error. I also pinged another user of our adapter, and this change would break them ([comment](https://github.com/go-delve/delve/pull/2617#issuecomment-885745255)) because as logic would suggest they expect success to mean success and act accordingly. 

> * VS Code only shows a modal dialog if the "showUser" property of the error message exists and is false. If "showUser" is missing, we would behave like before in order to preserve compatibility.

This would be great.  To clarify do you propose the following:
* showUser=true => show both modal dialog AND error regular pop-up?
* showUser=false => show neither
* no showUser => show only modal dialog?

**Comment by isidorn** [comment]
Since I do not have Go setup, @polinasok or @weinand can you please create a branch of Mock Debug that would behave like the go debugger in this case with respect to the `showUser` flag. That would make it easy for me to double check if all is good. Thanks!

**Comment by weinand** [comment]
@isidorn I will add this to mock-debug

**Comment by isidorn** [comment]
@weinand thanks a lot for updating mock debug, this was very helpful.
I have pushed a change as we agreed above.
@polinasok please try it out with vscode insiders from tomorrow and let me know how it goes. Thanks a lot!

**Comment by polinasok** [comment]
I tried this and can see that change in the behavior. I looks like it also impacts attach request failures as well. I am grateful for a quick turnaround and your responsiveness. However, I am concerned that this change will negatively impact your users. Because there was an automated pop-up for launch and attach errors, we chose to set showUser to false for all launch/attach errors to avoid double pop-ups that users would need to manage. With the new change, we loose the modal dialog for all errors, not just build errors. Therefore all errors that don't also log to Debug Console become silent. This, of course, can be addressed with a quick code change adding `showUser=true` as needed. But while we are anticipating this change, other adapter implementors that used the same logic might get caught off-guard.

**Comment by polinasok** [comment]
Oh, and another thought, what is the advantage of having both the modal and the bottom pop-up?
![image](https://user-images.githubusercontent.com/51177946/130191004-89c2bfb4-fc95-4617-8de5-a6a9ffe2e955.png)
The user must close both of them manually. Perhaps it makes more sense for `showUser` to replace the small bottom error with the visible modal dialog?

**Comment by isidorn** [comment]
@polinasok thanks for trying it out. All good feedback. Let's reopen so we try to address this.

1. I see. We could only have this logic for `launch` requests, however that sounds not consistent. What do you propose? If the `showUser` is not defined the error will still be shown as before. So I do not understand how this can catch adapter implementors off guard. Can you clarify pease?
2. When the dialog is shown we should not show the notification. Reason for this is that different code paths handle this and they are not in sync. I can look into this.

**Comment by weinand** [comment]
Yes, having a modal alert and a notification doesn't make sense.

**Comment by isidorn** [comment]
I have pushed a fix for 2.
However let's leave this open to see @polinasok response for 1.

**Comment by polinasok** [comment]
Clarification on 1. What I meant is that there might be implementers out there who chose explicit `showUser=false` not because they did not want to show anything to the user, but because they wanted to show the error only once. Since the editor always reliably triggered the visible modal alert, regardless of what the spec says, the `showUser` property was not controlling the overall visibility, but only the visibility of the duplicate notification. This fix changes the status quo for `false` values.

Specifically for Go, we will now need to change the adapter code to set `showUser=true` for all launch/attach errors other than the long build errors that go into Debug Console. Since the users might update their editor and their tools at different times, some might end up with duplicate notifications (old version of vscode + new version of adapter) and some users might end up with no notifications (new version of vscode + old version of adapter) for some short period of time. Eventually the dust will settle and everybody will have a consistent experience.

**Comment by isidorn** [comment]
@polinasok thanks for clarifying. I agree with you that the dust will settle eventually, and in the meantime the inconsistency which we might have is ok. So closing this as I plan no further changes here. Thanks

**Comment by polinasok** [comment]
> I agree with you that the dust will settle eventually, and in the meantime the inconsistency which we might have is ok.

We will be ok because we know this is coming and can time our changes accordingly. Implementers that set `showUser=false` explicitly might be caught off-guard when their launch/attach error become silent. I suggest clearly documenting this in your release notes, so they can quickly figure out what is going on.

When will this change be live? It is marked with Aug milestone. Does that mean early September?

**Comment by isidorn** [comment]
@polinasok yes as you say this will be in the Stable August release, and in practice that means early September.
It is already live in Insiders.
